File: Class Representation/Composition/C++/Composition.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

class LineItems
{
private:
    string name;
    int quantity;
    double unitPrice;

public:
    LineItems(const string &name, const int &quantity, const double &unitPrice) : name(name), quantity(quantity), unitPrice(unitPrice) {}

    double getSubTotal() const
    {
        return quantity * unitPrice;
    }

    string getProductName() const
    {
        return name;
    }

    void describe() const
    {
        cout << name << " x" << quantity
             << " @ $" << unitPrice
             << " = $" << getSubTotal() << endl;
    }
};

class Order
{
private:
    string orderId;
    vector<LineItems> lineItems;

public:
    Order(const string &orderId) : orderId(orderId) {}

    void addItem(const string &product, int quantity, double unitPrice)
    {
        lineItems.emplace_back(product, quantity, unitPrice);
    }

    void removeItem(const string &product)
    {
        lineItems.erase(
            remove_if(lineItems.begin(), lineItems.end(),
                      [&](const LineItems &item)
                      {
                          return item.getProductName() == product;
                      }),
            lineItems.end());
    }

    double getTotal() const
    {
        double total = 0;
        for (const auto &item : lineItems)
        {
            total += item.getSubTotal();
        }
        return total;
    }

    void printReceipt() const
    {
        cout << "Order: " << orderId << endl;
        for (const auto &item : lineItems)
        {
            item.describe();
        }
        cout << "Total: $" << getTotal() << endl;
    }
};

int main()
{

    Order order("ORD-1001");
    order.addItem("Wireless Mouse", 2, 29.99);
    order.addItem("USB-C Cable", 3, 9.99);
    order.addItem("Laptop Stand", 1, 49.99);

    order.printReceipt();
    return 0;
}

File: Class Representation/Composition/Typescript/Composition.ts
class LineItem {
  private productName: string;
  private quantity: number;
  private unitPrice: number;

  constructor(name: string, quantity: number, unitPrice: number) {
    this.productName = name;
    this.quantity = quantity;
    this.unitPrice = unitPrice;
  }

  getSubTotal(): number {
    return this.unitPrice * this.quantity;
  }

  getProductName(): string {
    return this.productName;
  }

  describe(): void {
    console.log(
      `${this.productName} x ${this.quantity} @ $ ${this.unitPrice} = $ ${this.getSubTotal}`,
    );
  }
}

class Order {
  private orderId: string;
  private lineItems: LineItem[];

  constructor(orderId: string) {
    this.orderId = orderId;
    this.lineItems = [];
  }

  addItem(name: string, quantity: number, unitPrice: number): void {
    this.lineItems.push(new LineItem(name, quantity, unitPrice));
  }

  removeItem(product: string): void {
    this.lineItems = this.lineItems.filter(
      (item) => item.getProductName() !== product,
    );
  }

  getTotal(): number {
    return this.lineItems.reduce(
      (total, item) => total + item.getSubTotal(),
      0,
    );
  }

  printReceipt(): void {
    console.log(`Order: ${this.orderId}`);
    for (const item of this.lineItems) {
      item.describe();
    }
    console.log(`Total: $${this.getTotal().toFixed(2)}`);
  }
}

function main(): void {
  const order = new Order("ORD-1001");
  order.addItem("Wireless Mouse", 2, 29.99);
  order.addItem("USB-C Cable", 3, 9.99);
  order.addItem("Laptop Stand", 1, 49.99);

  order.printReceipt();
}

main();
